### PR TITLE
Document extension API

### PR DESCRIPTION
**Pull request by pitrou**

### CODE REVIEW

File: docs/source/extending/interval-example.rst [hunk 1]
```diff
@@ -0,0 +1,320 @@
...
+We must now teach Numba how to actually generated code and data for
+the new operations.
+
+
+Defining the data model for native intervals
+--------------------------------------------
+
+As a general rule, :term:`nopython mode` does not work on Python objects
+as they are generated by the CPython interpreter.  The representations
+used by the interpreter are far too efficient for fast native code.
```
Comment (seibert) [inline docs/source/extending/interval-example.rst]: "far too inefficient"

**Comment by datnamer** [comment]
@pitrou I want to use a scipy [distributions](https://github.com/scipy/scipy/blob/v0.17.0/scipy/stats/_continuous_distns.py) object within a nopython function/in the field of a JIT Class. 

Most of the objects inherit from a superclass. Is this something that the extension API can cope with?

How about something more complex, like a SM model: https://github.com/statsmodels/statsmodels/blob/master/statsmodels/nonparametric/kernel_regression.py ?

Some have  multiple layers of inheritance and composition.

**Comment by seibert** [comment]
The extension API is pretty low level, and quite literal.  It doesn't directly respect inheritance, but also doesn't require that you provide implementations of base class methods, and let's you customize code generation completely independent of the python implementations.

Basically, you extend Numba at the base level by mapping incoming Python objects to Numba types, then those Numba types drive the whole type inference and lowering process after that point.

**Comment by seibert** [comment]
BTW, I think these docs look great and will merge.  We can take another edit pass in a future PR if needed.

**Comment by datnamer** [comment]
@seibert Does that mean it would still work for these two usecases?  I read through the docs and there was an example about directly translated a struct, but maybe I should reread more carefully.

**Comment by seibert** [comment]
Yes and no.  Yes it would work, but it would not be automated in the way you are imagining.  The extension API is not for compiling external code.  It is for describing how external types should be represented and methods and functions should be implemented.  If you added support for a scipy distribution object via this mechanism, Numba would not directly reuse any of the scipy implementation of those functions / classes.  Numba would rely on the extension author to provide all those implementations.

For comparison: Numba's support for numpy and ndarray is entirely due to us reimplementing all of those functions.  We do not directly compile any of the Python code in NumPy.  (There are various reasons for this...)

**Comment by datnamer** [comment]
Ah, I understand now. Thanks for the explanation. 

I don't look forward to re-implementing all of scipy distributions... I thought there may have been a way to specify types at the "top" and let it all flow down. But then again, python code is usually too mutable for that to work I guess.

**Comment by seibert** [comment]
We want to close that gap, for sure.  Writing everything again is not something anyone wants to do.  :)  In the case of NumPy, we had to because:

1. Numba was (and is) still very experimental, and we can't expect any mature project to alter their implementations to suit us.
2. Many of these implementations use Python features that Numba doesn't support, and even if we did, would not produce the fastest code.

In the ideal future, Numba would be able to compile most Python implementations used in libraries like NumPy and SciPy, etc, but then it might be necessary to hand write some implementations to more fully take advantage of Numba.

**Comment by datnamer** [comment]
Gotcha. That is really interesting and an exciting future.

How do you see numba getting there? More language constructs or more flexible type inference (tracing, branching etc)

**Pull request merged**